documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,179 @@
+where the `data` key contains the GTFS-RT Message encoded as JSON.
+
+`<prefix>/trip_modifications/#{id}/trip_modification`
+```json5
+{
+  "data": {
+    # <GTFS-RT Trip Modification Message as JSON>
+  }
+  "meta": {
+    // this field will be removed once we give users the ability to activate and
+    // deactivate detours themselves.
+    //
+    // Currently, this indicates that this message should not be published to
+    // applications outside of the MBTA, and is for internal testing and
+    // development only.
+    "is_draft?": boolean
+  }
+}
+```
+`<prefix>/trip_modifications/#{id}/shape`
+```json5
+{
+  "data": {
+    # <GTFS-RT Shape Message as JSON>
+  }
+}
+```
+
+### Subscriber Outcomes
+With this Structure, _theoretically_, the only subscription that Transit Data
+should need to make is to the following topics.
+```
+<env_prefix>/trip_modifications/+/trip_modification
+<env_prefix>/trip_modifications/+/shape
+```
+OR, Using only [wildcard topics](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/#heading-avoid-subscribing-to-wildcards):
+```
+<env_prefix>/trip_modifications/+/+
+<env_prefix>/trip_modifications/#
+```
+
+## Consequences
+
+This topic hierarchy requires that Skate is ensuring that the open retained
+messages on MQTT are synced with our own internal SOT of which
+Trip Modifications are active.
+
+
+## Alternatives Considered
+
+1.  A single topic in which typed events are published, with or without
+    differential feeds in addition to the typed events.
+    (Detour A exists now, Detour B now applies to trip xyz, Detour C is
+    deactivated, etc).
+
+    It seemed like the complexity of managing retained messages and a bunch of
+    message "types" would be a bit more of a hassle instead of leaning into MQTT
+    topics, in addition to figuring out retention/"rehydration".
+
+2.  Topics split out by detour ID with retained messages, but instead of
+    deactivating with a 0-byte message, deactivating with an explicit "deactivated"
+    message.
+
+    Before we 100% knew how retained messages were removed, we considered
+    publishing "deactivation" messages to the topics which contained
+    Trip Modifications previously. Given MQTT has a deletion method already, we
+    are opting to use that.
+
+3.  No deactivating at all - we just stop adding trip ID's to a particular detour
+    (or remove them if they've already been added) once a detour doesn't apply to
+    future or current trips anymore.
+
+    Because we need to keep Trip Modifications up to date with the trips that
+    the modification applies to, we'll be publishing frequent updates depending
+    on how many trips out we're configured to say will be affected. We
+    theoretically could publish a Trip Modification with an empty
+    `selected_trips` field, but this would go against the
+    [GTFS spec of Required: Many, Cardinality: One](https://gtfs.org/realtime/reference/#message-selectedtrips)
```

Comment:
This might be a misunderstanding on my part, but I thought that a trip ID uniquely specified a trip. Like the run of the 39 departing Back Bay today at 4:35 is a specific trip ID, and no other trip has the same ID. 

If I'm right, then is there anything wrong with having a `TripModification` with a `selected_trip_ids` that includes a trip that happened in the past?

Not saying that we need to go that route by any stretch, but this alternative would mean that "`TripModification` where all of the `selected_trip_ids` are trips that already happened" === "Deactivated Detour".

Reply:
Trip Id's are not unique across time.
So if the service ID for a trip _doesn't_ change, then that trip will be marked as detoured in perpetuity.

https://gtfs.org/schedule/reference/#tripstxt

> ### Trips
> [...]
>
> You may also notice there is no date field. That is because in GTFS, a `trip` is not an individual trip as a rider would perceive it (i.e. “the Inbound 89 trip that occurs at 8:39am on October 27th, 2020”), but a trip that occurs once on every date defined for a given service. 
>
> – https://github.com/firestack/gtfs_onboarding.nix/blob/82c011138de121e8ee1c520cc56f39bca880b5ba/GTFS_Onboarding.sqlite3.ipynb#Trips